modules/ingest-service-impl/src/main/java/org/opencastproject/ingest/scanner/InboxScannerService.java

```diff
@@ -179,29 +176,31 @@ public synchronized void updated(Dictionary properties) throws ConfigurationExce
     final int maxThreads = NumberUtils.toInt(Objects.toString(properties.get(INBOX_THREADS), "1"));
     final int maxTries = NumberUtils.toInt(Objects.toString(properties.get(INBOX_TRIES), "3"));
     final int secondsBetweenTries = NumberUtils.toInt(Objects.toString(properties.get(INBOX_TRIES_BETWEEN_SEC), "300"));
-    final Option<SecurityContext> secCtx = getUserAndOrganization(securityService, orgDir, orgId, userDir, userId)
-            .bind(new Function<Tuple<User, Organization>, Option<SecurityContext>>() {
-              @Override
-              public Option<SecurityContext> apply(Tuple<User, Organization> a) {
-                return some(new SecurityContext(securityService, a.getB(), a.getA()));
-              }
-            });
-    // Only setup new inbox if security context could be aquired
-    if (secCtx.isSome()) {
-      // remove old file install configuration
-      fileInstallCfg.foreach(removeFileInstallCfg);
-      // set up new file install config
-      fileInstallCfg = some(configureFileInstall(cc.getBundleContext(), inbox, interval));
-      // create new scanner
-      Ingestor ingestor = new Ingestor(ingestService, secCtx.get(), workflowDefinition,
-              workflowConfig, mediaFlavor, inbox, maxThreads, seriesService, maxTries, secondsBetweenTries);
-      this.ingestor = some(ingestor);
-      new Thread(ingestor).start();
-      logger.info("Now watching inbox {}", inbox.getAbsolutePath());
-    } else {
-      logger.warn("Cannot create security context for user {}, organization {}. "
-              + "Either the organization or the user does not exist", userId, orgId);
+
+    var securityContext = getUserAndOrganization(securityService, orgDir, orgId, userDir, userId)
+            .map(a -> new SecurityContext(securityService, a.getB(), a.getA()));
+    while (securityContext.isEmpty()) {
+      logger.debug("Could not create security context for user {}, organization {}. "
+              + "Either the organization or the user does not exist (yet).", userId, orgId);
+        try {
+          Thread.sleep(5000);
+        } catch (InterruptedException e) {
+          return;
+        }
+      securityContext = getUserAndOrganization(securityService, orgDir, orgId, userDir, userId)
+              .map(a -> new SecurityContext(securityService, a.getB(), a.getA()));
```

Comment:
Instead of busy waiting (potentially forever if the org does not exists / is removed), you can register a listener to the `OrganizationDirectoryService#addOrganizationDirectoryListener(OrganizationDirectoryListener)`. Those will be called once an organization is added, removed, updated.

Not sure if it is best to register the `InboxScannerService` instance or an anonymous in place implementation.

Reply:
Using `addOrganizationDirectoryListener()` is potentially possible but makes this horribly complex since you would have to deal with lots of concurrency problems that can happen. After all, the organization can already exist, the organization may only exist in the future, or it may happen exactly parallel to registering any listeners.

In fact, there is actually a good chance of this happening in parallel since organization creation and this code running roughly happens at the same time, which is the reason for this issue only popping up from time to time.

That is why you would need to register the listener, then check if the organization was already there, unregister it again if it existed already and would need a synchronized method to start the Ingestor to prevent starting it twice by accident. Given that this should usually only wait for a very short time, if at all, I don't think that complexity is worth it, do you?